File: app/src/main/java/com/example/pshash/Algorithm.hs
module Algorithm where

import Data.Char (ord, chr)

-- ┌───────────────────────────┐
-- │ GENERAL-PURPOSE FUNCTIONS │
-- └───────────────────────────┘

factorial :: Integer -> Integer
factorial n = product [1..n]

factorial' :: Integer -> Integer -> Integer
factorial' n m = product [(n-m+1)..n]

cnk :: Integer -> Integer -> Integer
cnk n k = div (factorial' n k) (factorial k)

length' :: [a] -> Integer
length' = toInteger . length

take' :: Integer -> [a] -> [a]
take' 0 _ = []
take' _ [] = []
take' n (a:as) = a : take' (n-1) as

insertAt :: a -> Integer -> [a] -> [a]
insertAt a 0 lst = a:lst
insertAt a n (l:lst) = l : insertAt a (n-1) lst
insertAt _ _ _ = []

break' :: (Eq a) => a -> [a] -> ([a], [a])
break' _ [] = ([],[])
break' a (a':rest)
  | a == a' = ([], rest)
  | otherwise = let (res1, res2) = break' a rest in (a' : res1, res2)

dropElementInfo :: ([a], Integer) -> (Integer, Integer)
dropElementInfo (src, m) = (length' src, m)

shiftString :: Integer -> String -> String
shiftString amt = map (chr . fromInteger . (`mod` 128) . (+ amt) . toInteger . ord)

class Shifting a where
  shift :: a -> Integer

instance (Shifting a) => Shifting [a] where
  shift = sum . map shift

instance Shifting Char where
  shift = toInteger . ord

combineHashing :: (a -> Integer -> b) -> (b -> Integer -> c) -> (a -> Integer -> Integer -> c)
combineHashing f g a = g . f a

mapHashing :: (Shifting b) => (a -> Integer -> b) -> (a -> Integer) -> ([a] -> Integer -> [b])
mapHashing _ _ [] _ = []
mapHashing f spr (a : as) key = b : mapHashing f spr as nextKey
  where
    (keyDiv, keyMod) = divMod key $ spr a
    b = f a keyMod
    nextKey = keyDiv + shift b

composeHashing :: (Shifting b) => (a -> Integer -> b) -> (a -> Integer) -> (b -> Integer -> c) -> (a -> Integer -> c)
composeHashing f spr g a key = g b nextKey
  where
    (keyDiv, keyMod) = divMod key $ spr a
    b = f a keyMod
    nextKey = keyDiv + shift b

-- ┌─────────────────────────────────────────────────────┐
-- │ PRE-DEFINED STRINGS FROM WHICH HASHES WILL BE DRAWN │
-- └─────────────────────────────────────────────────────┘

sourceLower :: [Char]
sourceLower = "ckapzfitqdxnwehrolmbyvsujg"

sourceUpper :: [Char]
sourceUpper = "RQLIANBKJYVWPTEMCZSFDOGUHX"

sourceSpecial :: [Char]
sourceSpecial = "=!*@?$%#&-+^"

sourceNumbers :: [Char]
sourceNumbers = "1952074386"

defaultConfiguration :: [([Char], Integer)]
defaultConfiguration = [(sourceLower, 8), (sourceUpper, 8), (sourceSpecial, 5), (sourceNumbers, 4)]

mediumConfiguration :: [([Char], Integer)]
mediumConfiguration = [(sourceLower, 5), (sourceUpper, 5), (sourceSpecial, 5), (sourceNumbers, 5)]

shortConfiguration :: [([Char], Integer)]
shortConfiguration = [(sourceLower, 4), (sourceUpper, 4), (sourceSpecial, 4), (sourceNumbers, 4)]

anlongConfiguration :: [([Char], Integer)]
anlongConfiguration = [(sourceLower, 7), (sourceUpper, 7), (sourceNumbers, 7)]

anshortConfiguration :: [([Char], Integer)]
anshortConfiguration = [(sourceLower, 4), (sourceUpper, 4), (sourceNumbers, 4)]

pinCodeConfiguration :: [([Char], Integer)]
pinCodeConfiguration = [(sourceNumbers, 4)]

mediumPinCodeConfiguration :: [([Char], Integer)]
mediumPinCodeConfiguration = [(sourceNumbers, 6)]

longPinCodeConfiguration :: [([Char], Integer)]
longPinCodeConfiguration = [(sourceNumbers, 8)]

-- ┌───────────────────────────┐
-- │ HASH GENERATING FUNCTIONS │
-- └───────────────────────────┘

chooseOrdered :: (Eq a, Shifting a) => ([a], Integer) -> Integer -> [a]
chooseOrdered (_, 0) _ = []
chooseOrdered ([], _) _ = []
chooseOrdered (src, m) key = curElt : chooseOrdered (filter (/= curElt) src, m - 1) nextKey
  where
    (keyDiv, keyMod) = divMod key $ length' src
    curElt = src !! fromIntegral keyMod
    nextKey = keyDiv + shift curElt

shuffleList :: (Eq a, Shifting a) => [a] -> Integer -> [a]
shuffleList src = chooseOrdered (src, length' src)

chooseOrderedSpread :: (Integer, Integer) -> Integer
chooseOrderedSpread (n, m) = factorial' n m

chooseOrderedSpread' :: ([a], Integer) -> Integer
chooseOrderedSpread' = chooseOrderedSpread . dropElementInfo

mergeTwoLists :: (Shifting a) => ([a], [a]) -> Integer -> [a]
mergeTwoLists ([], lst2) _ = lst2
mergeTwoLists (lst1, []) _ = lst1
mergeTwoLists (elt1:rest1, elt2:rest2) key
  | curKey < spr1 = elt1 : mergeTwoLists (rest1, elt2:rest2) (curKey + shift elt1)
  | otherwise = elt2 : mergeTwoLists (elt1:rest1, rest2) (curKey - spr1 + shift elt2)
  where
    spr1 = mergeTwoListsSpread (length' rest1, length' rest2 + 1)
    spr2 = mergeTwoListsSpread (length' rest1 + 1, length' rest2)
    curKey = mod key (spr1 + spr2)

mergeTwoListsSpread :: (Integer, Integer) -> Integer
mergeTwoListsSpread (m1, m2) = div (factorial (m1 + m2)) (factorial m1 * factorial m2)

mergeTwoListsSpread' :: ([a], [a]) -> Integer
mergeTwoListsSpread' (lst1, lst2) = mergeTwoListsSpread (length' lst1, length' lst2)

mergeLists :: (Shifting a) => [[a]] -> Integer -> [a]
mergeLists [] _ = []
mergeLists [l] _ = l
mergeLists [l1, l2] key = mergeTwoLists (l1, l2) key
mergeLists (l:ls) key = mergeTwoLists (l, mergeLists ls keyMod) nextKey
  where
    (keyDiv, keyMod) = divMod key $ mergeListsSpread $ map length' ls
    nextKey = keyDiv + shift l

mergeListsSpread :: [Integer] -> Integer
mergeListsSpread amts = div (factorial $ sum amts) (product $ map factorial amts)

mergeListsSpread' :: [[a]] -> Integer
mergeListsSpread' = mergeListsSpread . map length'

chooseAndMerge :: (Eq a, Shifting a) => [([a], Integer)] -> Integer -> [a]
chooseAndMerge = composeHashing
  (mapHashing chooseOrdered chooseOrderedSpread')
  (product . map chooseOrderedSpread')
  mergeLists

chooseAndMergeSpread :: [(Integer, Integer)] -> Integer
chooseAndMergeSpread amts = (product . map chooseOrderedSpread) amts * (mergeListsSpread . map snd) amts

chooseAndMergeSpread' :: [([a], Integer)] -> Integer
chooseAndMergeSpread' = chooseAndMergeSpread . map dropElementInfo

getHash :: (Eq a, Shifting a) => [([a], Integer)] -> Integer -> Integer -> [a]
getHash = combineHashing chooseAndMerge shuffleList

File: app/src/main/java/com/example/pshash/Algorithm.kt
package com.example.pshash
import java.math.BigInteger

fun tbi (n : Int) : BigInteger {
    return BigInteger.valueOf(n.toLong())
}

val bigZero = tbi(0)
val bigOne = tbi(1)

fun factorial (n : BigInteger) : BigInteger {
    return if (n == bigZero) tbi(1) else n * factorial(n - bigOne)
}

fun relativeFactorial (n : BigInteger, m : BigInteger) : BigInteger {
    return if (m == bigZero) bigOne else n * relativeFactorial(n - bigOne, m - bigOne)
}

fun shift (c : Char): BigInteger {
    return tbi(c.code)
}

fun shift2 (lst : List<Char>) : BigInteger {
    return lst.fold(bigZero) { acc, c ->
        acc + shift(c)
    }
}

fun shift3 (lst : List<List<Char>>) : BigInteger {
    return lst.fold(bigZero) { acc, c ->
        acc + shift2(c)
    }
}

val combineHashing : (
        (List<Pair<List<Char>, Int>>, BigInteger) -> List<Char>,
        (List<Char>, BigInteger) -> List<Char>
) -> (List<Pair<List<Char>, Int>>, BigInteger, BigInteger) -> List<Char> = { f, g ->
    { config, k1, k2 ->
        g(f(config, k1), k2)
    }
}

fun mapHashing (
    f : (Pair<List<Char>, Int>, BigInteger) -> List<Char>,
    spr : (Pair<List<Char>, Int>) -> BigInteger
) : (List<Pair<List<Char>, Int>>, BigInteger) -> List<List<Char>> {
    return { lst, key ->
        if (lst.isEmpty()) emptyList() else {
            val a : Pair<List<Char>, Int> = lst.first()
            val s = spr(a)
            val keyMod : BigInteger = key % s
            val keyDiv : BigInteger = key / s
            val b : List<Char> = f(a, keyMod)
            val prev : List<List<Char>> = mapHashing(f, spr)(lst.drop(1), keyDiv + shift2(b))
            listOf(b) + prev
        }
    }
}

fun composeHashing (
    f : (List<Pair<List<Char>, Int>>, BigInteger) -> List<List<Char>>,
    spr : (List<Pair<List<Char>, Int>>) -> BigInteger,
    g : (List<List<Char>>, BigInteger) -> List<Char>
) : (List<Pair<List<Char>, Int>>, BigInteger) -> List<Char> {
    return { a, key ->
        val s = spr(a)
        val keyMod = key % s
        val keyDiv = key / s
        val b = f(a, keyMod)
        g(b, keyDiv + shift3(b))
    }
}

val sourceLower : List<Char> = "ckapzfitqdxnwehrolmbyvsujg".toList()
val sourceUpper : List<Char> = "RQLIANBKJYVWPTEMCZSFDOGUHX".toList()
val sourceSpecial : List<Char> = "=!*@?$%#&-+^".toList()
val sourceNumbers : List<Char> = "1952074386".toList()

val defaultConfiguration : List<Pair<List<Char>, Int>> =
    listOf(Pair(sourceLower, 8), Pair(sourceUpper, 8), Pair(sourceSpecial, 5), Pair(sourceNumbers, 4))
val mediumConfiguration : List<Pair<List<Char>, Int>> =
    listOf(Pair(sourceLower, 5), Pair(sourceUpper, 5), Pair(sourceSpecial, 5), Pair(sourceNumbers, 5))
val shortConfiguration : List<Pair<List<Char>, Int>> =
    listOf(Pair(sourceLower, 4), Pair(sourceUpper, 4), Pair(sourceSpecial, 4), Pair(sourceNumbers, 4))
val anlongConfiguration : List<Pair<List<Char>, Int>> =
    listOf(Pair(sourceLower, 7), Pair(sourceUpper, 7), Pair(sourceNumbers, 7))
val anshortConfiguration : List<Pair<List<Char>, Int>> =
    listOf(Pair(sourceLower, 4), Pair(sourceUpper, 4), Pair(sourceNumbers, 4))
val pinCodeConfiguration : List<Pair<List<Char>, Int>> =
    listOf(Pair(sourceNumbers, 4))
val mediumPinCodeConfiguration : List<Pair<List<Char>, Int>> =
    listOf(Pair(sourceNumbers, 6))
val longPinCodeConfiguration : List<Pair<List<Char>, Int>> =
    listOf(Pair(sourceNumbers, 8))

fun chooseOrdered (
    src : Pair<List<Char>, Int>,
    key : BigInteger
) : List<Char> {
    if (src.second == 0 || src.first.isEmpty()) return emptyList()
    val len : BigInteger = tbi(src.first.size)
    val keyMod : Int = (key % len).toInt()
    val keyDiv : BigInteger = key / len
    val curElt : Char = src.first[keyMod]
    val prev : List<Char> = chooseOrdered(Pair(src.first.filter { c -> c != curElt }, src.second-1), keyDiv + shift(curElt))
    return listOf(curElt) + prev
}

val shuffleList : (
    List<Char>,
    BigInteger
) -> List<Char> = { lst, key ->
    chooseOrdered(Pair(lst, lst.size), key)
}

val chooseOrderedSpread : (
    Pair<List<Char>, Int>
) -> BigInteger = { src ->
    relativeFactorial(tbi(src.first.size), tbi(src.second))
}

val chooseOrderedSpreadMapped : (
    List<Pair<List<Char>, Int>>
) -> BigInteger = { srcs ->
    srcs.map { src -> chooseOrderedSpread(src) }.fold(bigOne, { acc, cur -> acc * cur})
}

val mergeTwoListsSpread : (
    BigInteger,
    BigInteger
) -> BigInteger = { m1, m2 ->
    factorial(m1 + m2) / (factorial(m1) * factorial(m2))
}

fun mergeTwoLists (
    src : Pair<List<Char>, List<Char>>,
    key : BigInteger
) : List<Char> {
    if (src.first.isEmpty()) return src.second
    if (src.second.isEmpty()) return src.first
    val elt1 : Char = src.first.first()
    val elt2 : Char = src.second.first()
    val spr1 : BigInteger = mergeTwoListsSpread(tbi(src.first.size-1), tbi(src.second.size))
    val spr2 : BigInteger = mergeTwoListsSpread(tbi(src.first.size), tbi(src.second.size-1))
    val curKey : BigInteger = key % (spr1 + spr2)
    if (curKey < spr1) {
        val prev : List<Char> = mergeTwoLists(Pair(src.first.drop(1), src.second), curKey + shift(elt1))
        return listOf(elt1) + prev
    } else {
        val prev : List<Char> = mergeTwoLists(Pair(src.first, src.second.drop(1)), curKey - spr1 + shift(elt2))
        return listOf(elt2) + prev
    }
}

fun mergeListsSpread (
    srcs : List<List<Char>>
) : BigInteger {
    var sum = 0
    var prod : BigInteger = bigOne
    for (src in srcs) {
        sum += src.size
        prod *= factorial(tbi(src.size))
    }
    return factorial(tbi(sum)) / prod
}

fun mergeLists (
    srcs : List<List<Char>>,
    key : BigInteger
) : List<Char> {
    if (srcs.isEmpty()) return emptyList()
    if (srcs.size == 1) return srcs[0]
    if (srcs.size == 2) return mergeTwoLists(Pair(srcs[0], srcs[1]), key)
    val l = srcs[0]
    val newSrcs = srcs.drop(1)
    val spr = mergeListsSpread(newSrcs)
    val keyMod = key % spr
    val keyDiv = key / spr
    return mergeTwoLists(Pair(l, mergeLists(newSrcs, keyMod)), keyDiv + shift2(l))
}

val chooseAndMerge : (
    List<Pair<List<Char>, Int>>,
    BigInteger
) -> List<Char> = composeHashing(
    mapHashing(
        { x, y -> chooseOrdered(x, y) },
        chooseOrderedSpread
    ),
    chooseOrderedSpreadMapped
) { x, y -> mergeLists(x, y) }

val getHash : (
    srcs : List<Pair<List<Char>, Int>>,
    BigInteger,
    BigInteger
) -> List<Char> = combineHashing(chooseAndMerge, shuffleList)